test/functorch/test_vmap.py

```diff
@@ -5152,6 +5161,145 @@ def check_gte_0(t):
+        nt = self._create_nt([4, None, 3], device=device)
+        self._vmap_test(f, (nt,))
+
+    @allowVmapFallbackUsage
+    def test_fallback_binary(self, device):
+        def f(x, y):
+            return x @ y
+
+        x = self._create_nt([5, None, 3], device=device)
+        y = self._create_nt([5, 3, None], device=device)
+        self._vmap_test(f, (x, y))
+
+    @allowVmapFallbackUsage
+    def test_fallback_binary_nt_and_unbatched_dense(self, device):
+        def f(x, y):
+            return x @ y
+
+        x = self._create_nt([5, None, 3], device=device)
+        y = torch.randn(3, 4, device=device)
+        self._vmap_test(f, (x, y), in_dims=(0, None))
+
+    @allowVmapFallbackUsage
+    def test_fallback_binary_nt_and_batched_dense(self, device):
+        def f(x, y):
+            return x @ y
+
+        x = self._create_nt([5, None, 3], device=device)
+        y = torch.randn(5, 3, 4, device=device)
+        self._vmap_test(f, (x, y))
+
+    def test_cat_batching_rule(self, device):
+        def f(x, y, dim):
+            return torch.cat([x, y], dim=dim)
+
+        # Different nested structure, same other dims
+        x = self._create_nt([3, None, 2], device=device)
+        y = self._create_nt([3, None, 2], device=device)
+        self._vmap_test(functools.partial(f, dim=0), (x, y))
+
+        x = self._create_nt([3, 2, None], device=device)
+        y = self._create_nt([3, 2, None], device=device)
+        self._vmap_test(functools.partial(f, dim=1), (x, y))
+
+        # Same nested structure, different other dims
+        x = self._create_nt([3, 2, None], device=device)
+        y = self._nt_from_similar(x, [-1, 4, -1])
+        self._vmap_test(functools.partial(f, dim=0), (x, y))
+
+        x = self._create_nt([3, None, 2], device=device)
+        y = self._nt_from_similar(x, [-1, -1, 4])
+        self._vmap_test(functools.partial(f, dim=1), (x, y))
+
+    # .shape calls don't work on NTs
+    # TODO: Fix this somehow?
+    @unittest.expectedFailure
+    def test_shape_call(self, device):
+        def f(x):
+            x.shape[0]
+            return x
+
+        x = self._create_nt([3, None, 2])
+        self._vmap_test(f, (x,))
+
+    def test_nt_with_nonzero_bdim_raises(self, device):
+        def f(x):
+            return x
+
+        x = self._create_nt([3, None, 2], device=device)
+        with self.assertRaisesRegex(
+                RuntimeError, "Nested tensors can only be vmapped over dim=0"):
+            vmap(f, in_dims=2)(x)
+
+    @allowVmapFallbackUsage
+    def test_fallback_with_nt_and_batched_dense_with_nonzero_bdim_raises(self, device):
+        def f(x, y):
+            return x @ y
+
+        x = self._create_nt([5, None, 3], device=device)
+        y = torch.randn(3, 5, 4, device=device)
+
+        with self.assertRaisesRegex(
+            RuntimeError,
+            "Fallback not supported for mixed nested / non-nested arguments without bdim=0"
+        ):
+            vmap(f, in_dims=(0, 1))(x, y)
```

Comment:
could we check that vmap with out_dims = "something that is not 0" errors out gracefully? That would usually trigger a transpose which is not supported for nested tensors.

Reply:
sure, note that `test_nt_with_nonzero_bdim_raises` covers the `in_dims` case so I just need one for `out_dims`